### ISSUE TITLE
Clarify "graphics commands"

### ISSUE THREAD

**Issue by ShabbyX** [KhronosGroup/Vulkan-Docs#1322]
The description of `vkCmdBindPipeline` says:

> Once bound, a pipeline binding affects subsequent graphics or compute commands in the command buffer until a different pipeline is bound to the bind point. The pipeline bound to VK_PIPELINE_BIND_POINT_COMPUTE controls the behavior of vkCmdDispatch and vkCmdDispatchIndirect. The pipeline bound to VK_PIPELINE_BIND_POINT_GRAPHICS controls the behavior of all drawing commands.

In the first sentence, it talks about "graphics or compute commands". The second part talks about `vkCmdDispatch*` and `vkCmdDraw*` calls (otherwise called _drawing commands_).

I cannot find a definition for "graphics commands". It is unclear whether graphics commands includes `vkCmdBeginRenderPass`. See here for a discussion on how this affects Broadcom: https://bugs.chromium.org/p/angleproject/issues/detail?id=4746

In summary, the situation in the bug is as follows:

- `vkCmdBeginCommandBuffer`
- `vkCmdBeginRenderPass` <- has multisampled attachments
- `vkCmdBindPipeline` <- is a multisampled pipeline
- `vkCmdEndRenderPass`
- `vkCmdBeginRenderPass` <- has singlesampled attachments, loadOp = CLEAR
- `vkCmdEndRenderPass`
- ...

The claim by Broadcom is that the multisampled pipeline that's bound in the first render pass affects "all subsequent graphics commands", including `vkCmdBeginRenderPass`, but that creates an invalid situation because the attachments being cleared are not multisampled.

I believe "graphics commands" to mean "drawing commands", but this is not defined anywhere. Please clarify.

**Comment by krOoze** [comment]
"graphics commands" likely means drawing and raytracing commands. But to me, it does not seem to matter. The sentence including the term is introductory, and more explicit descriptions follow, like:

> The pipeline bound to `VK_PIPELINE_BIND_POINT_GRAPHICS` controls the behavior of all _drawing commands_. [...] No other commands are affected by the pipeline state.

**Comment by ShabbyX** [comment]
I had missed that last sentence: _No other commands are affected by the pipeline state._ (I stopped reading after ray tracing!).

I agree that `vkCmdBeginRenderPass` shouldn't be affected by the currently bound pipeline, so I'll let Broadcom continue with the discussion if there are still doubts.

**Comment by krOoze** [comment]
I mean, `vkCmdBeginRenderPass` is for the most part a state command. I don't really see how the pipeline would affect it in any way even if it could. I mean, you can call a `vkCmdBeginRenderPass` without any pipeline bound at all (there is no pipeline VU).

I didn't read the linked bug thread (nor do I want to, TBH). Could you sum up here what behavior of `vkCmdBeginRenderPass` they think is changed by the bound pipeline?

**Comment by ShabbyX** [comment]
@krOoze that was my logic as well.

Let me repost the scenario that's troublesome for Broadcom:

- `vkCmdBeginCommandBuffer`
- `vkCmdBeginRenderPass` <- has multisampled attachments
- `vkCmdBindPipeline` <- is a multisampled pipeline
- `vkCmdEndRenderPass`
- `vkCmdBeginRenderPass` <- has singlesampled attachments, loadOp = CLEAR
- `vkCmdEndRenderPass`
- ...

The problem as they explained is:

> In a tiled system, the number of tiles representing the framebuffer will be different if the GraphicsPipeline is multisample or not. Loading and writing to/from a tile which is configured for a multisample pipeline doesn't have the same constraints if the attachment is multisample or not.

My understanding is that they decide how to tile the framebuffer based on the last bound pipeline's sample count, though it sounds to me like they should really be looking at the renderpass's attachments instead.

**Comment by gary-sweet** [comment]
We (Broadcom) have been looking at this again today. We have found a way to workaround the problem in our driver, so this may now be a moot point. The problem we see is when LOAD_OP_CLEAR is used to clear a single-sample attachment, but a multi-sample pipeline is bound. If there were any draw calls in that render pass, that would not be valid usage. The fact that the LOAD_OP_CLEAR doesn't count as a draw allows this as valid usage. We do now agree on that point. I can find nothing in the spec that disallows this.

It's worth pointing out that simply calling vkCmdClearAttachments() instead might have been more appropriate to achieve the same result.

**Comment by ShabbyX** [comment]
@gary-sweet Thank you for the update. Regarding `vkCmdClearAttachments()`, bear in mind that we are a GL driver with no information of what's coming next from the application. So for example we won't know if the render pass will end up being empty by the time we start it. But we digress.

> If there were any draw calls in that render pass, that would not be valid usage.

That's right, and we make sure there is another `vkCmdBindPipeline` with the appropriate sample count before issuing a draw.

**Comment by ShabbyX** [comment]
While the issue with Broadcom is resolved now, I'd leave this bug open as I still think it's helpful if Vulkan either avoided or defined "graphics commands" (and "compute commands", even if that has a smaller chance of confusion).

**Comment by sfricke-samsung** [comment]
Assigned to myself, going to look into improving spec language here for these cases

**Comment by oddhack** [comment]
This should be addressed in the 1.2.152 spec update.
